File: 79-word-search/word-search.cpp
class Solution {
public:
    int n, m;

  bool func(vector<vector<char> > & v, int i, int j, string& s, int ind)
 {
    if(ind == s.length())
    return true;

    if(i<0|| j<0 || i>= v.size()||  j >= v[0].size()|| s[ind] != v[i][j])
    return false;

    bool ans = false;

    char x = v[i][j];
    v[i][j] = ' ';

    // Check all four possible directions (down, up, right, left)
    ans |= func(v, i+1 , j , s , ind+1);
    ans |= func(v, i-1 , j , s , ind+1);
    ans |= func(v, i , j+1 , s , ind+1);
    ans |= func(v, i , j-1 , s , ind+1);

    v[i][j]= x ;
    return ans;

 }
    

    bool exist(vector<vector<char>>& board, string word) {
        n = board.size();
        m = board[0].size();
        
        if (n * m < word.size()) return false;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == word[0]) {
                   if(func(board, i, j , word , 0))
                    return true;
                }
            }
        }
        return false;
    }
};

File: 79-word-search/word-search.java
class Solution {

    boolean helper(char[][] board, String word, int i, int j, int len, int c, int n, int m, int[][] visited) {

        if (c == len) {
            return true;
        }

        if (i + 1 < n && visited[i + 1][j] == 0 && board[i + 1][j] == word.charAt(c)) {
            visited[i + 1][j] = 1;
            if (helper(board, word, i + 1, j, len, c + 1, n, m, visited))
                return true;
            visited[i + 1][j] = 0;
        }

        if (j + 1 < m && visited[i][j + 1] == 0 && board[i][j + 1] == word.charAt(c)) {
            visited[i][j + 1] = 1;
            if (helper(board, word, i, j + 1, len, c + 1, n, m, visited))
                return true;
            visited[i][j + 1] = 0;
        }

        if (i - 1 >= 0 && visited[i - 1][j] == 0 && board[i - 1][j] == word.charAt(c)) {
            visited[i - 1][j] = 1;
            if (helper(board, word, i - 1, j, len, c + 1, n, m, visited))
                return true;
            visited[i - 1][j] = 0;
        }

        if (j - 1 >= 0 && visited[i][j - 1] == 0 && board[i][j - 1] == word.charAt(c)) {
            visited[i][j - 1] = 1;
            if (helper(board, word, i, j - 1, len, c + 1, n, m, visited))
                return true;
            visited[i][j - 1] = 0;
        }

        return false;
    }

    public boolean exist(char[][] board, String word) {

        int n = board.length;
        int m = board[0].length;
        int len = word.length();

        int[][] visited = new int[n][m];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {

                if (board[i][j] == word.charAt(0)) {

                    visited[i][j] = 1;

                    if (helper(board, word, i, j, len, 1, n, m, visited))
                        return true;

                    visited[i][j] = 0;
                }
            }
        }

        return false;
    }
}